index.js

```diff
@@ -1,35 +1,65 @@
-/*!
- * forwarded
- * Copyright(c) 2014 Douglas Christopher Wilson
- * MIT Licensed
- */
-
-/**
- * Module exports.
- */
+'use strict'
 
-module.exports = forwarded
+var Processor = require('./lib/processor')
+var schemas = require('./lib/schemas')
 
 /**
  * Get all addresses in the request, using the `X-Forwarded-For` header.
  *
- * @param {Object} req
+ * @param {http.IncomingMessage} req
  * @api public
  */
 
-function forwarded(req) {
+module.exports = function forwarded (req, options) {
+  options = options || {}
+
+  var opts = {
+    // default to only common + standard
+    // array order matters here
+    schemas: options.schemas || [
+      'xff',
+      'rfc7239'
+    ]
+  }
+
+  // consistent case
+  opts.schemas.map(Function.prototype.call, String.prototype.toLowerCase)
+
   if (!req) {
     throw new TypeError('argument req is required')
   }
 
-  // simple header parsing
-  var proxyAddrs = (req.headers['x-forwarded-for'] || '')
-    .split(/ *, */)
-    .filter(Boolean)
-    .reverse()
-  var socketAddr = req.connection.remoteAddress
-  var addrs = [socketAddr].concat(proxyAddrs)
+  // start with default values from socket connection
+  var forwarded = {
+    addrs: [req.connection.remoteAddress],
+    by: null,
+    host: req.headers && req.headers.host ? req.headers.host : undefined,
+    port: req.connection.remotePort.toString(),
+    ports: [req.connection.remotePort.toString()],
+    proto: req.connection.encrypted ? 'https' : 'http'
```

Comment:
Are you sure this actually works in Node.js 0.6 and the others? I only ask because the test is just a mock, so we have to verify manually if we cannot add a real test.